inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -2657,43 +2661,63 @@ private void readAllBeanDefinitionClasses() {
                     if (!hasNonSingletonCandidate && !candidate.isSingleton()) {
                         hasNonSingletonCandidate = true;
                     }
+                    if (candidate.findAnnotation(Order.class).isPresent()) {
+                        hasOrderAnnotation = true;
+                    }
                     addCandidateToList(resolutionContext, beanType, candidate, beansOfTypeList, qualifier, candidateCount == 1);
                 }
                 if (!hasNonSingletonCandidate) {
                     allCandidatesAreSingleton = true;
                 }
-                beans = beansOfTypeList;
+                beanRegistrations = beansOfTypeList;
             } else {
                 if (LOG.isDebugEnabled()) {
                     LOG.debug("Found no possible candidate beans of type [{}] for qualifier: {} ", beanType.getName(), qualifier);
                 }
                 allCandidatesAreSingleton = true;
-                beans = Collections.emptySet();
+                beanRegistrations = Collections.emptySet();
             }
 
-            if (beans != Collections.EMPTY_SET) {
-                if (Ordered.class.isAssignableFrom(beanType)) {
-                    beans = beans.stream().sorted(OrderUtil.COMPARATOR).collect(StreamUtils.toImmutableCollection());
+            if (beanRegistrations != Collections.EMPTY_SET) {
+                if (Ordered.class.isAssignableFrom(beanType) || hasOrderAnnotation) {
+                    beanRegistrations = beanRegistrations.stream().sorted(BEAN_REGISTRATION_COMPARATOR).collect(StreamUtils.toImmutableCollection());
                 } else {
-                    beans = Collections.unmodifiableCollection(beans);
+                    beanRegistrations = Collections.unmodifiableCollection(beanRegistrations);
                 }
             }
 
+            Collection<T> beans = beanRegistrations.stream().flatMap(b -> Stream.of(b.getBean())).collect(StreamUtils.toImmutableCollection());
             if (allCandidatesAreSingleton) {
                 initializedObjectsByType.put(key, (Collection<Object>) beans);
             }
-            if (LOG.isDebugEnabled() && !beans.isEmpty()) {
+            if (LOG.isDebugEnabled() && !beanRegistrations.isEmpty()) {
```

Comment:
I'm not sure this change is correct, shouldn't it be checking the resolved beans?

Reply:
Yep you're correct.